webui/react/src/components/UPlot/UPlotChart.tsx

```diff
@@ -26,141 +25,208 @@ interface ScaleZoomData {
-          } ],
-          setScale: [ (uPlot: uPlot, scaleKey: string) => {
-            const currentMax = uPlot.posToVal(scaleKey === 'x' ? uPlot.bbox.width : 0, scaleKey);
-            const currentMin = uPlot.posToVal(scaleKey === 'x' ? 0 : uPlot.bbox.height, scaleKey);
-            let max = scalesZoomData.current[scaleKey]?.max;
-            let min = scalesZoomData.current[scaleKey]?.min;
-
-            if (max == null || currentMax > max) max = currentMax;
-            if (min == null || currentMin < min) min = currentMin;
-
-            scalesZoomData.current[scaleKey] = { isZoomed: isZoomed.current, max, min };
-
-            /*
-             * Save the scale info if zoomed in and clear it otherwise.
-             * This info will be used to restore the zoom when remounting
-             * the chart, which can be caused by new series data, chart option
-             * changes, etc.
-             */
-            if (!scalesRef.current) scalesRef.current = {};
-            if (isZoomed.current) {
-              scalesRef.current[scaleKey] = uPlot.scales[scaleKey];
-            } else {
-              delete scalesRef.current[scaleKey];
-            }
-            if (Object.keys(scalesRef.current).length === 0) scalesRef.current = undefined;
-          } ],
-        },
-        scales: scalesRef.current,
-        width: chartDivRef.current.offsetWidth,
+        width: chartDivRef.current?.offsetWidth,
       },
-      options,
+      options || {},
     ) as uPlot.Options;
 
-    const plotChart = new uPlot(optionsExtended, normalizedData, chartDivRef.current);
+  const optionsRef = useRef<uPlot.Options>(getAugmentedOptions(options));
 
-    return () => {
-      plotChart.destroy();
+  const [ hasData, normalizedData ] = useMemo(
+    () => getNormalizedData(data, optionsRef.current?.mode),
+    [ data ],
+  );
+
+  useEffect(() => {
+    if (!chartDivRef.current) return;
+    let newOptions = getAugmentedOptions(options);
+    if (!chartRef.current || shouldUpdate(optionsRef.current, newOptions, chartRef.current)) {
+      /**
+       * TODO: instead of returning true or false,
+       * shouldUpdate could return actions:
+       * 'noAction', 'resetSeries', and 'remount', etc.
+       * then the we could put addSeries and delSeries here
+       * or remount, etc.
+       * depending on the action returned
+       */
+      chartRef.current?.destroy();
       chartRef.current = undefined;
+
+      const isZoomed = Object.values(scalesZoomData.current).some((i) => i.isZoomed === true);
+      if (!isZoomed) {
+        optionsRef.current.scales = {};
+        scalesZoomData.current = {};
+      } else {
+        /** TODO: preserve zoom when new series is selected?
+       * may also want to preserve other user interactions with the chart
+       * by taking some options from chartRef.current and putting them in newOptions
+       */
+      }
+      newOptions = uPlot.assign(
+        optionsRef.current,
```

Comment:
we do assign like this to pick up any user interactions that caused the chart to update the options ref in the meantime.

Reply:
re pt 2: 

user interactions that we dont handle update the `optionsRef.current` and the chart itself without updating the `props.options` so in general we dont need to explicitly handle them, but we dont want to clobber them on update